README.md
[![Go report](https://goreportcard.com/badge/github.com/lcomrade/wallblog)](https://goreportcard.com/report/github.com/lcomrade/wallblog)
[![Docker Hub Pulls](https://img.shields.io/docker/pulls/lcomrade/wallblog)](https://hub.docker.com/r/lcomrade/wallblog)
[![Release](https://img.shields.io/github/v/release/lcomrade/wallblog)](https://github.com/lcomrade/wallblog/releases/latest)
[![License](https://img.shields.io/github/license/lcomrade/wallblog)](LICENSE)

**WallBlog** - lightweight blogging engine with markdown support.

Features:
- Page formatting in markdown and html
- Custom page header and footer (in markdown)
- Custom CSS style
- Automatic sitemap generation
- Custom error pages
- HTTP and HTTPS


## 📦Install binary release
### UNIX-like systems
Example for Linux amd64.
For other OS/arch change the link end.
```
curl --output /usr/local/bin/wallblog -L https://github.com/lcomrade/wallblog/releases/latest/download/wallblog.linux.amd64
chmod 755 /usr/local/bin/wallblog
```

### Docker container
Read more on the [Docker Hub page](https://hub.docker.com/r/lcomrade/wallblog).



## 🔨Build from source
Build deps:
- Git
- GNU Make
- Golang

```bash
git clone https://github.com/lcomrade/wallblog.git
cd ./wallblog
make
```

You can find result of build in `./dist/` directory.



## ⚙️Configuration
### `/etc/wallblog/config.json`
To apply the changes in this file, you must restart the program.

Default:
```json
{
	"WebRoot": "/var/lib/wallblog",
	"HTTP": {
		"Enable": true,
		"Port": ":80"
	},
	"HTTPS": {
		"Enable": false,
		"Port": ":443",
		"Cert": "",
		"Key": ""
	},
	"Overwrite": {
		"Host": "",
		"Protocol": ""
	},
	"SiteMap": {
		"Enable": true,
		"URL": "/sitemap.xml",
		"SkipHidden": true
	},
	"Page": {
		"AutoTitle": {
			"Enable": true,
			"Prefix": "",
			"Sufix": ""
		},
		"AddToHead": [],
		"TemplateMode": {
			"EnableForMainPage": true,
			"EnableForPagePart": true
		}
	}
}
```


### `/var/lib/wallblog/*`
`**.md` files - normal Markdown with support for HTML tags.

`**.htmlp` files - only HTML tags are supported.
Put there what is usually between `<body></body>`.

`**` files - serve according to their MIME type.

#### Pages and files
- `**/index.htmlp` or `**/index.md` - works like `index.html`

#### Custom page design
- `./article_start.htmlp` or `./article_start.md` - custom beginning of the article
- `./article_end.htmlp` or `./article_end.md` - custom end of article
- `./header.htmlp` or `./header.md` - page header
- `./footer.htmlp` or `./footer.md` - page footer
- `./style.css` - page CSS style
- `./favicon.ico` - site icon

#### Custom error pages
- `./error.css` - error page CSS style
- `./403.htmlp` or `./403.md`
- `./404.htmlp` or `./404.md`
- `./500_permission_denied.htmlp` or `./500_permission_denied.md`
- `./500_file_read_timeout.htmlp` or `./500_file_read_timeout.md`
- `./500_unknown.htmlp` or `./500_unknown.md`



## 📑Documentation
- [Article/Error pages](docs/page.md)
- [Config: Page Head](docs/page_head.md)
- [Config: Site Map](docs/sitemap.md)
- [Config: Reverse proxy](docs/reverse_proxy.md)
- [FAQ](docs/faq.md)
- [Changelog](CHANGELOG.md)
- [MD2HTML: Markdown Syntax Guide](https://github.com/lcomrade/md2html/blob/main/docs/syntax_guide.md)


CHANGELOG.md
# Changelog
Semantic versioning is used (https://semver.org/).

## v1.3.2
- Added `article_start` and `article_end`
- Added build in variables (`{{URL.Full}}` and `{{URL.Path}}`)
- Disallow access to config files. Now return 403 HTTP code.
- Fix: site URL detection
- Update Golang 1.15.15 => 1.18.1

## v1.3.1
- Fix: HTML tags in autotitle

## v1.3.0
- Added automatic page title generation
- Allowed to add additional tags to `<head>...</head>`

## v1.2.0
- Added automatic sitemap generation
- Added reverse proxy support
- Allow use `index.htmlp` as index page
- Documented sitemap, reverse proxy and FAQ

## v1.1.0
- Added custom error pages
- Added support for HTMLP format
- Page rendering is documented

## v1.0.0
This is the first release with only basic functionality implemented here:
- Page formatting in Markdown
- Custom page header and footer (in Markdown)
- Custom CSS style
- HTTP and HTTPS


internal/handler/handler.go
// Copyright (C) 2022 Leonid Maslakov.

// This file is part of wallblog.

// wallblog is free software: you can redistribute it
// and/or modify it under the terms of the
// GNU Affero Public License as published by the
// Free Software Foundation, either version 3 of the License,
// or (at your option) any later version.

// wallblog is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Affero Public License for more details.

// You should have received a copy of the GNU Affero Public License along with wallblog.
// If not, see <https://www.gnu.org/licenses/>.

package handler

import (
	"errors"
	"github.com/lcomrade/wallblog/internal/cfg"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

var errHttpCode403 = errors.New("http code: 403 Forbidden")

var Config cfg.Config

var noAccessURLs = []string{
	"/article_start.htmlp", "article_start.md",
	"/article_end.htmlp", "article_end.md",
	"/header.htmlp", "/header.md",
	"/footer.htmlp", "/footer.md",
	"/403.htmlp", "/403.md",
	"/404.htmlp", "/404.md",
	"/500_permission_denied.htmlp", "/500_permission_denied.md",
	"/500_file_read_timeout.htmlp", "/500_file_read_timeout.md",
	"/500_unknown.htmlp", "/500_unknown.md",
}

func Hand(rw http.ResponseWriter, req *http.Request) {
	if req.Host != "" {
		req.URL.Host = req.Host
	}

	// Sanitize URL
	splitOrigURL := strings.Split(req.URL.Path, "/")
	req.URL.Path = "/"

	for _, part := range splitOrigURL {
		if part == "." {
			continue
		}

		if part == ".." {
			continue
		}

		if req.URL.Path != "/" {
			req.URL.Path = req.URL.Path + "/" + part

		} else {
			req.URL.Path = "/" + part
		}
	}

	// Find path
	url := req.URL.Path
	path := filepath.Join(Config.WebRoot, url)
	ext := filepath.Ext(path)

	// Skip URLs
	for _, skipURL := range noAccessURLs {
		if url == skipURL {
			errWrite(errHttpCode403, rw, req)
			return
		}
	}

	// Redirect
	if strings.HasSuffix(url, "/index.html") {
		urlRune := []rune(url)
		newURL := string(urlRune[0:len(urlRune)-11]) + "/"
		redirectWrite(newURL, rw, req)
		return
	}

	if strings.HasSuffix(url, "/index.htmlp") {
		urlRune := []rune(url)
		newURL := string(urlRune[0:len(urlRune)-12]) + "/"
		redirectWrite(newURL, rw, req)
		return
	}

	if strings.HasSuffix(url, "/index.md") {
		urlRune := []rune(url)
		newURL := string(urlRune[0:len(urlRune)-9]) + "/"
		redirectWrite(newURL, rw, req)
		return
	}

	// Get file info
	file, err := os.Stat(path)
	if err != nil {
		errWrite(err, rw, req)
		return
	}

	// Directory
	if file.IsDir() {
		dirHand(rw, req, path)
		return
	}

	// *.md file
	if ext == ".md" {
		mdHand(rw, req, path)
		return
	}

	// *.htmlp file
	if ext == ".htmlp" {
		htmlpHand(rw, req, path)
		return
	}

	// Other files
	fileHand(rw, req, path)
}
